{
  "url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704",
  "repository_url": "https://api.github.com/repos/quarkusio/quarkus",
  "labels_url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704/labels{/name}",
  "comments_url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704/comments",
  "events_url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704/events",
  "html_url": "https://github.com/quarkusio/quarkus/issues/28704",
  "id": 1416066822,
  "node_id": "I_kwDOCFbutM5UZ3cG",
  "number": 28704,
  "title": "Broken Redis client for cluster mode (MOVED Exception)",
  "labels": [
    {
      "id": 985346620,
      "node_id": "MDU6TGFiZWw5ODUzNDY2MjA=",
      "url": "https://api.github.com/repos/quarkusio/quarkus/labels/kind/bug",
      "name": "kind/bug",
      "color": "d73a4a",
      "default": false,
      "description": "Something isn't working"
    },
    {
      "id": 2184343143,
      "node_id": "MDU6TGFiZWwyMTg0MzQzMTQz",
      "url": "https://api.github.com/repos/quarkusio/quarkus/labels/area/redis",
      "name": "area/redis",
      "color": "0366d6",
      "default": false,
      "description": ""
    }
  ],
  "state": "closed",
  "locked": false,
  "milestone": {
    "url": "https://api.github.com/repos/quarkusio/quarkus/milestones/280",
    "html_url": "https://github.com/quarkusio/quarkus/milestone/280",
    "labels_url": "https://api.github.com/repos/quarkusio/quarkus/milestones/280/labels",
    "id": 10034258,
    "node_id": "MI_kwDOCFbutM4AmRxS",
    "number": 280,
    "title": "3.6.0.CR1",
    "description": "",
    "creator": {
      "login": "gsmet",
      "id": 1279749,
      "node_id": "MDQ6VXNlcjEyNzk3NDk=",
      "avatar_url": "https://avatars.githubusercontent.com/u/1279749?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/gsmet",
      "html_url": "https://github.com/gsmet",
      "followers_url": "https://api.github.com/users/gsmet/followers",
      "following_url": "https://api.github.com/users/gsmet/following{/other_user}",
      "gists_url": "https://api.github.com/users/gsmet/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/gsmet/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/gsmet/subscriptions",
      "organizations_url": "https://api.github.com/users/gsmet/orgs",
      "repos_url": "https://api.github.com/users/gsmet/repos",
      "events_url": "https://api.github.com/users/gsmet/events{/privacy}",
      "received_events_url": "https://api.github.com/users/gsmet/received_events",
      "type": "User",
      "site_admin": false
    },
    "open_issues": 0,
    "closed_issues": 185,
    "state": "closed",
    "created_at": "2023-10-11T07:34:36Z",
    "updated_at": "2024-01-14T09:44:21Z",
    "due_on": null,
    "closed_at": "2023-11-15T12:08:43Z"
  },
  "comments": 13,
  "created_at": "2022-10-20T06:50:25Z",
  "updated_at": "2023-10-19T17:46:11Z",
  "closed_at": "2023-10-19T17:46:06Z",
  "active_lock_reason": null,
  "body": "### Describe the bug\r\n\r\nHi,\r\n\r\nthere is an issue with the Redis client in cluster mode. Here is my config:\r\n```\r\nquarkus.redis.hosts=redis://localhost:7001,redis://localhost:7002,redis://localhost:7003,redis://localhost:7004,redis://localhost:7005,redis://localhost:7006\r\nquarkus.redis.client-type=cluster\r\nquarkus.redis.replicas=share\r\n```\r\nI have my cluster up and running in local for all those ports:\r\n```\r\nantony@pop-os:~$ redis-cli -c -p 7001\r\n127.0.0.1:7001> CLUSTER INFO\r\ncluster_state:ok\r\ncluster_slots_assigned:16384\r\ncluster_slots_ok:16384\r\ncluster_slots_pfail:0\r\ncluster_slots_fail:0\r\ncluster_known_nodes:6\r\ncluster_size:3\r\ncluster_current_epoch:6\r\ncluster_my_epoch:1\r\ncluster_stats_messages_ping_sent:107731\r\ncluster_stats_messages_pong_sent:107678\r\ncluster_stats_messages_sent:215409\r\ncluster_stats_messages_ping_received:107673\r\ncluster_stats_messages_pong_received:107731\r\ncluster_stats_messages_meet_received:5\r\ncluster_stats_messages_received:215409\r\ntotal_cluster_links_buffer_limit_exceeded:0\r\n...\r\nantony@pop-os:~$ redis-cli -c -p 7001\r\n127.0.0.1:7001> PING\r\nPONG\r\nantony@pop-os:~$ redis-cli -c -p 7002\r\n127.0.0.1:7001> PING\r\nPONG\r\nantony@pop-os:~$ redis-cli -c -p 7003\r\n127.0.0.1:7003> PING\r\nPONG\r\n...\r\n```\r\nI am running this simple code for testing:\r\n```java\r\n    // get cluster info and create a new field in a map\r\n    @Scheduled(every = \"4s\", delayed = \"3s\")\r\n    Uni<Void> test() {\r\n        System.out.println(\"\\nRunning test...\");\r\n        return redisClient.execute(\"CLUSTER\", \"INFO\")\r\n            .invoke(r -> System.out.println(\"\\nCluster Info --->\\n: \" + r))\r\n            .chain(() -> redisClient.execute(\"HSET\", \"myDict\", \"field-\" + UUID.randomUUID(), \"val\"))\r\n            .onFailure().recoverWithItem(e -> {\r\n                Log.error(\"Error returned: \", e);\r\n                return null;\r\n            })\r\n            .replaceWithVoid();\r\n    }\r\n```\r\nThis is the output I am getting:\r\n```\r\nCluster Info --->\r\n: cluster_state:ok\r\ncluster_slots_assigned:16384\r\ncluster_slots_ok:16384\r\ncluster_slots_pfail:0\r\n...\r\n\r\n2022-10-20 00:26:31,039 ERROR [com.cha.pau.scu.Test] (vert.x-eventloop-thread-1) Error returned: : MOVED 9564 127.0.0.1:7002\r\n```\r\nSo I do get the cluster info but any other command fails (I have tried multiple commands). I have tried with different profiles (dev, prod), with `quarkus dev` and running the jar; I always get the same error. After reading a bit it seems like this is just a redirect notification for the client to resend the request to the indicated node (https://redis.io/docs/reference/cluster-spec/), but the quarkus client doesn't do the redirect, it just throws the error. \r\n\r\n### Output of `java -version`\r\n\r\nopenjdk version \"11.0.16.1\" 2022-08-12 LTS\r\nOpenJDK Runtime Environment Corretto-11.0.16.9.1 (build 11.0.16.1+9-LTS)\r\nOpenJDK 64-Bit Server VM Corretto-11.0.16.9.1 (build 11.0.16.1+9-LTS, mixed mode)\r\n\r\n### Quarkus version or git rev\r\n\r\n2.13.0\r\n\r\n### Additional information\r\n\r\nRedis version:\r\n```\r\nantony@pop-os:~$ redis-server --version\r\nRedis server v=7.0.5 sha=00000000:0 malloc=jemalloc-5.2.1 bits=64 build=82289c7437798a30\r\nantony@pop-os:~$ redis-cli --version\r\nredis-cli 7.0.5\r\n```\r\nTo create the cluster I am simply running `create-cluster start` and `create-cluster create` (https://redis.io/docs/manual/scaling/#create-and-use-a-redis-cluster).\r\n\r\n##### UPDATE:\r\nJust run it with Redis version 6.2 (cluster mode) and I get the same error.",
  "reactions": {
    "url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704/reactions",
    "total_count": 0,
    "+1": 0,
    "-1": 0,
    "laugh": 0,
    "hooray": 0,
    "confused": 0,
    "heart": 0,
    "rocket": 0,
    "eyes": 0
  },
  "timeline_url": "https://api.github.com/repos/quarkusio/quarkus/issues/28704/timeline",
  "performed_via_github_app": null,
  "state_reason": "completed"
}
